common-lib/src/main/scala/com/gu/mediaservice/GridClient.scala

```diff
@@ -132,73 +121,88 @@ class GridClient(apiKey: String, services: Services, maxIdleConnections: Int, de
     }
   }
 
-  private def makeRequestAsync(url: URL, apiKey: String): Future[Response] = {
-    val request = new Request.Builder().url(url).header(apiKeyHeaderName, apiKey).build
-    val promise = Promise[Response]()
-    httpClient.newCall(request).enqueue(new Callback {
-      override def onFailure(call: Call, e: IOException): Unit = promise.failure(e)
 
-      override def onResponse(call: Call, response: Response): Unit = promise.success(response)
-    })
-    promise.future
-  }
-
-  private def validateResponse(res: ResponseWrapper, url: URL): Unit = {
-    import res._
-    if (statusCode != 200 && statusCode != 404) {
-      val errorMessage = s"breaking the circuit of full image projection, downstream API: $url is in a bad state, code: $statusCode"
-      val downstreamErrorMessage = res.bodyAsString
-
-      val errorJson = Json.obj(
-        "level" -> "ERROR",
-        "errorStatusCode" -> statusCode,
-        "message" -> Json.obj(
-          "errorMessage" -> errorMessage,
-          "downstreamErrorMessage" -> downstreamErrorMessage
-        )
-      )
-      logger.error(errorJson.toString())
-      throw new DownstreamApiInBadStateException(errorMessage, downstreamErrorMessage)
+  private def validateResponse[T](
+                                   response: WSResponse,
+                                   url: URL,
+                                   foundFn: ResponseWrapper => Option[T],
+                                   notFoundFn: ResponseWrapper => Option[T],
+                                   errorFn: Option[ResponseWrapper => Exception]
+                                 ): Option[T] = {
+    val res = processResponse(response, url)
+    res.statusCode match {
+      case 200 => foundFn(res)
+      case 404 => notFoundFn(res)
+      case failCode => errorFn match {
+        case Some(fn) => throw fn(res)
+        case _ =>
+          val errorMessage = s"breaking the circuit of full image projection, downstream API: $url is in a bad state, code: $failCode"
```

Comment:
This message isn't super relevant any more.

Reply:
Agreed.